packages/react/src/components/fileInput/FileInput.tsx

```diff
@@ -0,0 +1,554 @@
+      }
+    }
+  };
+
+  const handleMultipleChange = (files: File[]) => {
+    if (files.length > 0) {
+      const { validFiles, validationErrors } = runValidations(files);
+
+      const [replacedFiles, newFiles] = validFiles.reduce(
+        (acc: [File[], File[]], file: File) => {
+          if (findDuplicateByNameAndType(selectedFiles, file)) {
+            return [[...acc[0], file], acc[1]];
+          }
+          return [acc[0], [...acc[1], file]];
+        },
+        [[], []],
+      );
+
+      if (validationErrors.length > 0) {
+        setInvalidText(getValidationErrorsMessage(validationErrors, files.length));
+      }
+
+      if (validFiles.length > 0) {
+        const selectedWithoutReplacedFiles = selectedFiles.filter(
+          (selectedFile: File) => !findDuplicateByNameAndType(replacedFiles, selectedFile),
+        );
+        setSelectedFiles([...selectedWithoutReplacedFiles, ...replacedFiles, ...newFiles]);
+        setProcessSuccessText(getAddSuccessMessage(language, validFiles.length, files.length));
+      }
+    }
+  };
+
+  const onFilesChange = (files: File[]) => {
+    clearNotificationTexts();
+
+    if (multiple) {
+      handleMultipleChange(files);
+    } else {
+      handleSingleFileChange(files);
+    }
+  };
+
+  const onRemoveFileFromList = (fileToRemove: File, index: number) => {
+    clearNotificationTexts();
+
+    const selectedFilesWithoutRemoved = selectedFiles.filter(
+      (file: File) => !isEqualFileBy(['name', 'type', 'size', 'lastModified'], file, fileToRemove),
+    );
+    setSelectedFiles(selectedFilesWithoutRemoved);
+    setInputStateText(getRemoveSuccessMessage(language));
+
+    if (selectedFilesWithoutRemoved.length > 0) {
+      fileListFocusIndexRef.current = index > 0 ? index - 1 : 0;
+      setInputStateText(getRemoveSuccessMessage(language));
+    } else if (inputRef.current) {
+      inputRef.current.focus();
+      setInputStateText(getNoFilesAddedMessage(language));
+    }
+  };
+
+  const onDragEnter = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    event.stopPropagation();
+    setIsDragOverDrop(true);
+  };
+
+  const onDragLeave = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    event.stopPropagation();
+    setIsDragOverDrop(false);
+  };
+
+  useEffect(() => {
+    if (didMountRef.current && onChange) {
+      onChange(selectedFiles);
+    } else {
+      didMountRef.current = true;
+    }
+    // Clear input value on every change to ensure it triggers a onChange event when files are added
+    resetFileInputValue();
+  }, [selectedFiles, onChange]);
+
+  // Compose aria-describedby attribute
+  const ariaDescribedBy: string = [
+    composeAriaDescribedBy(id, helperTextToUse, errorTextToUse, successText, infoTextToUse),
+    hasFilesSelected && fileListId,
+  ]
+    .filter((text) => !!text)
+    .join(' ');
+
+  return (
+    <>
+      <InputWrapper {...wrapperProps}>
```

Comment:
Is the `InputWrapper` necessary here? There's an input, but it's hidden from the screen. If I recall right, the InputWrapper will bring all sorts of styles related to text input? Though now when we don't have a text input here, those styles should not match? Could it complicate or interfere with styles of the component otherwise? It might bring some unnecessary styles that might have unknown side effects hmm.. Though it does bring the needed labels and info texts. I guess if we want to reuse those this is the way to go.

Reply:
I have to think about this 🤔  It would mean duplicate code